### PR TITLE
Collect and expose runtime's image storage usage via Kubelet's /stats/summary endpoint

### PR DESCRIPTION
**Pull request by vishh**
This information is useful to users since docker images are typically not stored on the root filesystem.

Kubelet will also consume this feature in the future to decide is evicting images will help with disk usage on the nodes.

cc @kubernetes/sig-node

### CODE REVIEW

**Comment by vishh** [comment]
@timstclair: Can you do an initial review?

File: cluster/images/hyperkube/master.json [hunk 1]
```diff
@@ -31,7 +31,7 @@
               "/hyperkube",
               "apiserver",
               "--service-cluster-ip-range=10.0.0.1/24",
-              "--insecure-bind-address=127.0.0.1",
+              "--insecure-bind-address=0.0.0.0",
```
Comment (timstclair) [inline cluster/images/hyperkube/master.json]: This looks unrelated? Also, doesn't this expose the insecure outside the local machine? Intentional?

File: pkg/kubelet/api/v1alpha1/stats/types.go [hunk 2]
```diff
@@ -46,6 +46,16 @@ type NodeStats struct {
 	// Stats pertaining to total usage of filesystem resources on the rootfs used by node k8s components.
 	// NodeFs.Used is the total bytes used on the filesystem.
 	Fs *FsStats `json:"fs,omitempty"`
+	// Stats about the underlying container runtime.
+	Runtime *RuntimeStats `json:"runtime,omitempty"`
+}
+
+// Stats pertaining to the underlying container runtime.
+type RuntimeStats struct {
+	// Stats about the underlying filesystem where container images are stored.
```
Comment (timstclair) [inline pkg/kubelet/api/v1alpha1/stats/types.go]: Do we want to include a name to which runtime we're using here?

File: pkg/kubelet/dockertools/images.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+// imageStatsProvider exposes stats about all images currently available.
+type imageStatsProvider struct {
+	// Docker remote API client
+	c DockerInterface
+}
+
+func (isp *imageStatsProvider) ImageStats() (*runtime.ImageStats, error) {
+	images, err := isp.c.ListImages(docker.ListImagesOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list docker images - %v", err)
```
Comment (timstclair) [inline pkg/kubelet/dockertools/images.go]: (nit) we usually separate message & err with a colon `: %v`

File: pkg/kubelet/dockertools/images.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+	}
+	// A map of all the image layer to its corresponding size.
```
Comment (timstclair) [inline pkg/kubelet/dockertools/images.go]: (nit) plurals are off, I think you mean: `A map of image layers to their corresponding size.`

File: pkg/kubelet/dockertools/images.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+				continue
+			}
+			key := layer.ID
+			// Some of the layers are empty.
+			// We are hoping that these layers are unique to each image.
+			// Still keying with the CreatedBy field to be safe.
+			if key == "" || key == "<missing>" {
+				key = key + layer.CreatedBy
+			}
+			imageMap[key] = uint64(layer.Size)
```
Comment (timstclair) [inline pkg/kubelet/dockertools/images.go]: I'm assuming that the point of the map here is to dedup layers shared across images? If that's the case, should you verify the the sizes match?

File: pkg/kubelet/dockertools/images_test.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+	"github.com/stretchr/testify/assert"
+)
+
+func TestImageStatsNoImages(t *testing.T) {
+	fakeDockerClient := NewFakeDockerClientWithVersion("1.2.3", "1.2")
+	isp := &imageStatsProvider{fakeDockerClient}
+	st, err := isp.ImageStats()
+	as := assert.New(t)
+	as.NoError(err)
+	as.True(st.TotalStorageBytes == 0)
```
Comment (timstclair) [inline pkg/kubelet/dockertools/images_test.go]: (nit) prefer assert.Equal

File: pkg/kubelet/dockertools/images_test.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+		},
+		{
+			ID: "busybox-new",
+		},
+	})
+	isp := &imageStatsProvider{fakeDockerClient}
+	st, err := isp.ImageStats()
+	as := assert.New(t)
+	as.NoError(err)
+	as.Equal(uint64(1300), st.TotalStorageBytes, "expected %d, got %d", 1000, st.TotalStorageBytes)
```
Comment (timstclair) [inline pkg/kubelet/dockertools/images_test.go]: expected values are inconsistent (assign a constant).

File: pkg/kubelet/dockertools/images_test.go [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+	fakeHistoryData := map[string][]docker.ImageHistory{
+		"busybox": {
+			{
+				ID:        "0123456",
+				CreatedBy: "foo",
+				Size:      100,
+			},
+			{
+				ID:        "0123457",
+				CreatedBy: "bar",
```
Comment (timstclair) [inline pkg/kubelet/dockertools/images_test.go]: (nit) name this something like "duplicate1" to make the intention clear (same below)

File: pkg/kubelet/server/server.go [hunk 9]
```diff
@@ -105,9 +106,17 @@ func (a *filteringContainer) RegisteredHandlePaths() []string {
...
-func ListenAndServeKubeletServer(host HostInterface, resourceAnalyzer stats.ResourceAnalyzer, address net.IP, port uint, tlsOptions *TLSOptions, auth AuthInterface, enableDebuggingHandlers bool) {
+func ListenAndServeKubeletServer(
+	host HostInterface,
+	resourceAnalyzer stats.ResourceAnalyzer,
+	address net.IP,
+	port uint,
+	tlsOptions *TLSOptions,
+	auth AuthInterface,
+	enableDebuggingHandlers bool,
+	runtime kubecontainer.Runtime) {
```
Comment (timstclair) [inline pkg/kubelet/server/server.go]: Can you just expose `GetRuntime` in the HostInterface instead?

**Comment by timstclair** [comment]
Finished review pass.

File: pkg/kubelet/server/server.go (hunk 9 above)
Comment (vishh) [inline pkg/kubelet/server/server.go]: I'd like to inject dependencies explicitly instead of passing in an opaque
object. I dislike the way we currently handle dependencies.

**Comment by yujuhong** [comment]
> Kubelet will also consume this feature in the future to decide is evicting images will help with disk usage on the nodes.

Is the intention to look at the total size to see if it's even *worth* evicting some images? The size of all images won't help decide what image to evict.

File: cluster/images/hyperkube/master.json (hunk 1 above)
Comment (vishh) [inline cluster/images/hyperkube/master.json]: Nope. Reverting this..

File: pkg/kubelet/api/v1alpha1/stats/types.go (hunk 2 above)
Comment (vishh) [inline pkg/kubelet/api/v1alpha1/stats/types.go]: I don't think so. Our APIs abstract the runtime out (mostly)

File: pkg/kubelet/dockertools/images.go (hunk 3 above)
Comment (vishh) [inline pkg/kubelet/dockertools/images.go]: Is this applicable across the repo? Almost all the code I have written uses `-`....

File: pkg/kubelet/dockertools/images.go (hunk 4 above)
Comment (vishh) [inline pkg/kubelet/dockertools/images.go]: Fixing it.

File: pkg/kubelet/dockertools/images.go (hunk 5 above)
Comment (vishh) [inline pkg/kubelet/dockertools/images.go]: I don't think so. Layers IDs are expected to be unique, except if the ID is `<missing>`

File: pkg/kubelet/dockertools/images_test.go (hunk 6 above)
Comment (vishh) [inline pkg/kubelet/dockertools/images_test.go]: Fixing.

File: pkg/kubelet/dockertools/images_test.go (hunk 7 above)
Comment (vishh) [inline pkg/kubelet/dockertools/images_test.go]: Fixing it

**Comment by Random-Liu** [comment]
@yujuhong Thanks. This needs a little rebase after #23506 get merged.

File: pkg/kubelet/dockertools/images_test.go (hunk 8 above)
Comment (vishh) [inline pkg/kubelet/dockertools/images_test.go]: Good suggestion. Fixing it.

**Comment by vishh** [comment]
> Is the intention to look at the total size to see if it's even worth evicting some images? The size of all images won't help decide what image to evict.

Yes. In a separate PR, I plan on cleaning up the image GC logic as well.

**Comment by vishh** [comment]
@timstclair: Thanks for the review. Addressed all comments.

File: pkg/kubelet/dockertools/images.go (hunk 3 above)
Comment (timstclair) [inline pkg/kubelet/dockertools/images.go]: I don't think there's anything in the style guide about it, but:
```console
$ git grep -E '(glog|fmt)\.(Warning|Info|Error)f\(\"[^"].*:\ %v\".*\berr' | wc -l
1200
$ git grep -E '(glog|fmt)\.(Warning|Info|Error)f\(\"[^"].*\ -\ %v\".*\berr' | wc -l
25
```

**Comment by vishh** [comment]
@timstclair I rebased the patch and addressed comments. PTAL

File: pkg/kubelet/container/runtime.go [hunk 10]
```diff
@@ -86,6 +92,10 @@ type Runtime interface {
 	ListImages() ([]Image, error)
 	// Removes the specified image.
 	RemoveImage(image ImageSpec) error
+	// Returns Image statistics.
+	ImageStats() (*ImageStats, error)
+	// Returns statistics of all the images currently available.
+	//	ImageStats() (*ImageStats, error)
```
Comment (timstclair) [inline pkg/kubelet/container/runtime.go]: Delete this?

Comment (vishh) [inline pkg/kubelet/container/runtime.go]: Done

File: pkg/kubelet/dockertools/images.go (hunk 5 above)
Comment (timstclair) [inline pkg/kubelet/dockertools/images.go]: If that's the case, why not just get rid of the imageMap and keys, and just do a cumulative sum?

File: pkg/kubelet/server/stats/summary.go [hunk 11]
```diff
@@ -83,8 +85,11 @@ func (sp *summaryProviderImpl) Get() (*stats.Summary, error) {
 	if err != nil {
 		return nil, err
 	}
-
-	sb := &summaryBuilder{sp.fsResourceAnalyzer, node, nodeConfig, rootFsInfo, imageFsInfo, infos}
+	imageStats, err := sp.runtime.ImageStats()
+	if err != nil || imageStats == nil {
+		return nil, err
```
Comment (timstclair) [inline pkg/kubelet/server/stats/summary.go]: nit: maybe it would be better to log the error, and try to keep going in this case?

**Comment by timstclair** [comment]
Just 2 more nits...

**Comment by derekwaynecarr** [comment]
Other than nits, this Lgtm

File: pkg/kubelet/dockertools/images.go (hunk 5 above)
Comment (vishh) [inline pkg/kubelet/dockertools/images.go]: Yeah. That makes sense. I'm updating the code now.

File: pkg/kubelet/server/stats/summary.go (hunk 11 above)
Comment (vishh) [inline pkg/kubelet/server/stats/summary.go]: If we are going down that path, I'd do it for all the stats. Or we can identify the ones that are required and ignore the optional ones. WDYT?
In any case, we can change that in a separate PR.

**Comment by vishh** [comment]
@timstclair Addressed your nits. PTAL

**Comment by yifan-gu** [comment]
cc @euank @sjpotter

**Comment by vishh** [comment]
@timstclair Just did a rebase. PTAL when you get a chance.

File: pkg/kubelet/server/stats/summary.go (hunk 11 above)
Comment (timstclair) [inline pkg/kubelet/server/stats/summary.go]: I agree, but also lets tackle this in a separate PR.

File: pkg/kubelet/dockertools/images.go [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+			// Some of the layers are empty.
+			// We are hoping that these layers are unique to each image.
+			// Still keying with the CreatedBy field to be safe.
+			if key == "" || key == "<missing>" {
+				key = key + layer.CreatedBy
+			}
+			if !imageMap.Has(key) {
+				ret.TotalStorageBytes += uint64(layer.Size)
+			}
+			imageMap.Insert(key)
```
Comment (timstclair) [inline pkg/kubelet/dockertools/images.go]: Doesn't this mean we'll only add the first layer for each image? Is that intentional?

Comment (timstclair) [inline pkg/kubelet/dockertools/images.go]: Do the unit tests still pass with this? If it's not intentional and the tests do still pass, please update the tests to fail on this.

Comment (vishh) [inline pkg/kubelet/dockertools/images.go]: Hmm. `history` contains all the layers for an image. Let's look at an example:

```shell
$ docker history bef5c7a893fd
IMAGE               CREATED             CREATED BY                                      SIZE                COMMENT
bef5c7a893fd        4 days ago          /bin/sh -c #(nop) ENTRYPOINT ["/usr/bin/cadvi   0 B                 
5714946567e8        4 days ago          /bin/sh -c #(nop) EXPOSE 8080/tcp               0 B                 
d5092c6dfc49        4 days ago          /bin/sh -c #(nop) ADD file:5047bc1f57482bfdc9   23.75 MB            
fb1790e81294        4 days ago          /bin/sh -c apk add --update ca-certificates d   19.99 MB            
3d239f12c85f        4 weeks ago         /bin/sh -c #(nop) ENV GLIBC_VERSION=2.23-r1     0 B                 
86fbae24586c        4 weeks ago         /bin/sh -c #(nop) MAINTAINER dengnan@google.c   0 B                 
342c0650e86b        7 weeks ago         /bin/sh -c #(nop) ADD file:cda4b589f22e7984e3   5.258 MB 
```

For this example, history would be a list of layer elements where each line in the output above matches a layer.
The key for each layer will be mostly unique, expect for the `<missing>` case, which is handled by using the `Created By` field. That field is expected to contain the command that was run for that layer and hence it must be unique.

Am I misreading my code?

Comment (timstclair) [inline pkg/kubelet/dockertools/images.go]: Ah, sorry, my mistake. I misread `key` as the image key, not the layer key. What you have makes sense.

**Comment by timstclair** [comment]
LGTM, needs rebase.

**Comment by vishh** [comment]
Just did a rebase. I'm adding the lgtm label again. FYI @timstclair

**Comment by Random-Liu** [comment]
Just FYI, the perfdash shows that this PR increases the runtime CPU usage by about 0.05 cores (from 0.3 cores to 0.35 cores).
It is not significant, but it would be good to track it here so that in the future if we want to do some performance optimization, we can revisit this. :)

**Comment by yujuhong** [comment]
> Just FYI, the perfdash shows that this PR increases the runtime CPU usage by about 0.05 cores (from 0.3 cores to 0.35 cores).

What build was this? I couldn't find it.

**Comment by Random-Liu** [comment]
@yujuhong [kubernetes-e2e-gce-serial 1155](http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gce-serial/1155/)

**Comment by yujuhong** [comment]
IIUC, the PR lists the images and invoke `ImageHistory` call for each image. We should be able to cache the output, and only call `ImageHistory` for images we haven't seen before.

**Comment by Random-Liu** [comment]
@yujuhong Agree. We could file an issue for this, :)
